### PR TITLE
Enhancing iceberg integration.

### PR DESCRIPTION
**Pull request by zhljen**
Update to iceberg 0.5.0 with metadata refresh retry limitation.
Enhance iceberg table loading with metrics for monitoring.
Add table check to prevent bad table loading.

### CODE REVIEW

File: metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableHandler.java [hunk 1]
```diff
@@ -36,59 +36,85 @@
...
+                result = this.icebergTableOpWrapper.getPartitionMetricsMap(icebergTable, null);
             }
         } catch (ParseException ex) {
             throw new MetacatBadRequestException("Iceberg filter parse error");
         }
-        return
-            ScanSummary.of(icebergTable.newScan())  //the top x records
-                .limit(connectorContext.getConfig().getIcebergTableSummaryFetchSize())
-                .build();
+        final long duration = registry.clock().wallTime() - start;
```
Comment (ajoymajumdar) [inline metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableHandler.java]: If the above logic fails, the metrics will not be reported.

File: metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableHandler.java [hunk 2]
```diff
@@ -99,7 +125,15 @@ public IcebergTableUtil(final ConnectorContext connectorContext) {
      * @return iceberg table
      */
     public Table getIcebergTable(final QualifiedName tableName, final String tableMetadataLocation) {
-        return new IcebergMetastoreTables(tableMetadataLocation).load(tableName.toString());
+        this.icebergTableCriteria.checkCriteria(tableName, tableMetadataLocation);
+        final long start = this.registry.clock().wallTime();
+        log.debug("Loading icebergTable {} from {}", tableName, tableMetadataLocation);
+        final Table table = new IcebergMetastoreTables(tableMetadataLocation).load(tableName.toString());
+        final long duration = registry.clock().wallTime() - start;
+        log.info("Time taken to getIcebergTable {} is {} ms", tableName, duration);
```
Comment (ajoymajumdar) [inline metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableHandler.java]: Same here. If the above logic fails, the metrics will not be reported.

File: metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableOpWrapper.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.Map;
+
+/**
+ * Iceberg table operation wrapper.
+ *
+ * @author zhenl
+ * @since 1.2.0
+ */
+@AllArgsConstructor
+public class IcebergTableOpWrapper {
```
Comment (ajoymajumdar) [inline metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableOpWrapper.java]: What was the reason to create this class?

File: metacat-connector-hive/src/test/groovy/com/netflix/metacat/connector/hive/HiveConnectorFastPartitionSpec.groovy [hunk 4]
```diff
@@ -49,122 +48,51 @@ class HiveConnectorFastPartitionSpec extends Specification {
     @Shared
     DirectSqlSavePartition directSqlSavePartition = Mock(DirectSqlSavePartition)
     @Shared
-    ConnectorRequestContext connectorContext = new ConnectorRequestContext(1, null);
```
Comment (ajoymajumdar) [inline metacat-connector-hive/src/test/groovy/com/netflix/metacat/connector/hive/HiveConnectorFastPartitionSpec.groovy]: Did we remove test cases?

File: metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/DataMetadataMetricConstants.java [hunk 5]
```diff
@@ -23,7 +23,7 @@
  * @author zhenl
  * @since 1.2.0
  */
-public final class DataMetricConstants {
+public final class DataMetadataMetricConstants {
```
[the remark this answers is not in the document]
Comment (zhljen) [reply metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/DataMetadataMetricConstants.java]: The name is a bit strange, but it has to be named as 'DataMetadata' vs 'Metadata' to separate from 'userMetadata'..

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/properties/HiveProperties.java [hunk 6]
```diff
@@ -98,5 +98,9 @@
     public static class Iceberg {
         private boolean enabled;
         private int fetchSizeInTableSummary = 100;
+        /* each retry needs a s3 access, default to 0 as no retry*/
```
[the remark this answers is not in the document]
Comment (zhljen) [reply metacat-common-server/src/main/java/com/netflix/metacat/common/server/properties/HiveProperties.java]: sure

File: metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergRequestMetrics.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    enum Type {
+        counter,
+        gauge,
+        timer
+    }
+
+    private final String metricName;
+
+    IcebergRequestMetrics(final IcebergRequestMetrics.Type type, final String measure) {
+        this.metricName = "metacat.iceberg." + type.name() + "." + type.name() + "." + measure;
```
[the remark this answers is not in the document]
Comment (zhljen) [reply metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergRequestMetrics.java]: sure.

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/properties/HiveProperties.java [hunk 8]
```diff
@@ -98,5 +98,9 @@
...
+        private int refreshFromMetadataLocationRetryNumber;
```
[the remark this answers is not in the document]
Comment (zhljen) [reply metacat-common-server/src/main/java/com/netflix/metacat/common/server/properties/HiveProperties.java]: The default is 0, which is explicit.

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/properties/HiveProperties.java [hunk 9]
```diff
@@ -98,5 +98,9 @@
     public static class Iceberg {
```
[the remark this answers is not in the document]
Comment (zhljen) [reply metacat-common-server/src/main/java/com/netflix/metacat/common/server/properties/HiveProperties.java]: To allow extending from our internal part.

[the remark this answers is not in the document]
Comment (zhljen) [reply metacat-common-server/src/main/java/com/netflix/metacat/common/server/properties/HiveProperties.java]: To allow extending from the internal part.

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/properties/HiveProperties.java [hunk 10]
```diff
@@ -98,5 +98,9 @@
...
         private boolean enabled;
         private int fetchSizeInTableSummary = 100;
+        /* each retry needs a s3 access, default to 0 as no retry*/
+        private int refreshFromMetadataLocationRetryNumber;
+        /* loading metadata consumes memory, cap to 50m as default */
+        private long maxMetadataFileSize = 50 * 1024 * 1024; //50m
```
[the remark this answers is not in the document]
Comment (zhljen) [reply metacat-common-server/src/main/java/com/netflix/metacat/common/server/properties/HiveProperties.java]: will do.

File: metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableHandler.java [hunk 11]
```diff
@@ -99,7 +125,15 @@ public IcebergTableUtil(final ConnectorContext connectorContext) {
      * @return iceberg table
      */
     public Table getIcebergTable(final QualifiedName tableName, final String tableMetadataLocation) {
-        return new IcebergMetastoreTables(tableMetadataLocation).load(tableName.toString());
+        this.icebergTableCriteria.checkCriteria(tableName, tableMetadataLocation);
```
[the remark this answers is not in the document]
Comment (zhljen) [reply metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableHandler.java]: that's correct.

File: metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableHandler.java (hunk 1 above)
Comment (zhljen) [reply metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableHandler.java]: Will move to finally block.

File: metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableHandler.java (hunk 2 above)
Comment (zhljen) [reply metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableHandler.java]: Will move to finally block.

File: metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableOpWrapper.java (hunk 3 above)
Comment (zhljen) [reply metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/IcebergTableOpWrapper.java]: To abstract out the iceberg operation. Especially, for unit testing. Mockup deep down into the iceberg library inside is too involved which should be abstracted out.

File: metacat-connector-hive/src/test/groovy/com/netflix/metacat/connector/hive/HiveConnectorFastPartitionSpec.groovy (hunk 4 above)
Comment (zhljen) [reply metacat-connector-hive/src/test/groovy/com/netflix/metacat/connector/hive/HiveConnectorFastPartitionSpec.groovy]: Nope, this change is to consolidate them into one test.

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/properties/DefaultConfigImpl.java [hunk 12]
```diff
@@ -453,4 +453,21 @@ public int getIcebergTableSummaryFetchSize() {
...
+    @Override
+    public int getIcebergRefreshFromMetadataLocationRetryNumber() {
+        return this.metacatProperties.getHive().getIceberg().getRefreshFromMetadataLocationRetryNumber();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public long getIcebergMaxMetadataFileSize() {
```
[the remark this answers is not in the document]
Comment (zhljen) [reply metacat-common-server/src/main/java/com/netflix/metacat/common/server/properties/DefaultConfigImpl.java]: Internal implementation will use this configuration to valid the iceberg metadata file size.

File: metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/DataMetadataMetricConstants.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (anupom) [reply metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/iceberg/DataMetadataMetricConstants.java]: Makes sense, thanks

**Pull request merged**